keps/sig-storage/3476-volume-group/README.md

```diff
@@ -0,0 +1,1647 @@
+
+* Some applications may not want to use ApplicationSnapshot CRD because they don’t use Kubernetes workload APIs such as StatefulSet, Deployment, etc. Instead, they have developed their own operators. In this case it is more convenient to use VolumeGroup to manage persistent volumes used in those applications.
+
+* Application quiesce is time consuming. Some users may not want to do application quiesce very frequently for that reason. For example, a user may want to run weekly backups with application quiesce and nightly backups without application quiesce but with consistency group support which provides crash consistency across all volumes in the group.
+
+### Future use cases
+
+* VolumeGroup can be used to manage group replication or consistency group replication if the storage system supports it. Note replication is out of scope for this proposal. It is mentioned here as a potential future use case.
+
+* VolumeGroup can be used to manage volume placement to either spread the volumes across storage pools or stack the volumes on the same storage pool. Related KEPs proposing the concept of storage pool for volume placement is as follows:
+  https://github.com/kubernetes/enhancements/pull/1353
+  https://github.com/kubernetes/enhancements/pull/1347
+We may not really need a VolumeGroup for this use case. A StoragePool is probably enough. This is to be determined.
+
+### Goals
+
+* Provide an API to manage multiple volumes together in a group.
+* Provide an API to take a snapshot of a group of volumes.
+* The group API should be generic and extensible so that it may be used to support other features in the future.
+
+### Non Goals
+
+* A VolumeGroup may potentially be used to support group replication in the future, but providing design on replication group is not in the scope of this KEP. This can be discussed in the future.
+* Provide a design to facilitate volume placement using the group API (To be determined).
+
+## Proposal for VolumeGroup and VolumeGroupSnapshot
+
+This proposal introduces new CRDs VolumeGroup, VolumeGroupContent, VolumeGroupClass, VolumeGroupSnapshot, VolumeGroupSnapshotContent, and VolumeGroupSnapshotClass.
+
+### Create VolumeGroup
+
+Create new VolumeGroup can be done in several ways:
+
+Phase 1 (Note: only Phase 1 will be covered in this KEP which is targeting Alpha in K8s v1.26):
+1. Create an empty group first, then create a new PVC with the group name. This will create a new volume and add that volume to the already created group. When deleting this volume group, all volumes in the group will be deleted together with the group. A CSI driver supporting CREATE_DELETE_VOLUME_GROUP controller capability MUST implement this feature.
+2. Create an empty group first, then add an existing PVC to the group one by one. A CSI driver supporting VOLUME_GROUP_ADD_REMOVE_EXISTING_VOLUME MUST implement this feature.
```

Comment:
Do you know if CSI drivers can only support one but not the other? For example, I think CSI drivers that can support 2) can also support 1). Can we simplify and only support 2)?

Reply:
I do know CSI drivers that can only support (1) but not (2).  Typically block type storage systems can support both, but shared filesystems type storage systems can only support (1) but not (2).